add update-unwanted-dependencies.sh to track unwanted dependencies

pacoxu opened the pull request:
#### What type of PR is this?
/kind feature
/area code-organization
/sig architecture

#### What this PR does / why we need it:

****Verify**: Run ./hack/lint-dependencies.sh**
- check forbidden_repos, exit if has,  the user should update it in unwanted-dependencies.json
- check unwanted-dependencies, exit if not has,  and the user should update it in unwanted-dependencies.json

Test results:
```
➜  kubernetes git:(track-unwanted-dependencies-1) hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
2021/09/18 13:32:05 The following dependencies **are not allowed:**
github.com/hashicorp/golang-lru
2021/09/18 13:32:05 !!! Please update status in ./hack/unwanted-dependencies.json
exit status 1

➜  kubernetes git:(track-unwanted-dependencies-1) ✗ hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
2021/09/18 13:33:36 The following dependencies **are removed:**
k8s.io/klog
2021/09/18 13:33:36 !!! Please update status in ./hack/unwanted-dependencies.json
exit status 1

➜  kubernetes git:(track-unwanted-dependencies-1) ✗ hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
2021/09/18 13:33:19 The following dependencies are removed:
k8s.io/klog
2021/09/18 13:33:19 The following dependencies are not allowed:
k8s.io/klog/v2
2021/09/18 13:33:19 !!! Please update status in ./hack/unwanted-dependencies.json
exit status 1

➜  kubernetes git:(track-unwanted-dependencies-1) hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
All pinned versions of checked dependencies match their preferred version.

```

#### Which issue(s) this PR fixes:
Fixes #102145

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @liggitt @gautierdelorme  @dims

pacoxu commented:
A quick discussion with Gautier on slack, some explanation on the PR for track

My initial idea:
> For  forbidden repo, that’s a choice to adding them in the forbidden_repos
> # List of dependencies we need to avoid dragging back into kubernetes/kubernetes 
>  forbidden_repos=( 
>    "k8s.io/klog"  # we have switched to klog v2, so avoid klog v1 
>  ) 
> One choice: Keep the list here.
> Another choice: Or add the hack/.not_allowed_dependencies  file and read the file here.

@gautierdelorme 
> to me hack/.not_allowed_dependencies seems more manageable but I don't really mind
> telling go add the module to hack/.not_allowed_dependencies seems simpler than go edit hack/lint-dependencies.sh and add it to forbidden_repos

pacoxu commented on hack/.unwanted_dependencies:
```diff
@@ -0,0 +1,6 @@
+github.com/go-openapi/spec
+github.com/go-openapi/strfmt
+github.com/go-openapi/validate
+go.mongodb.org/mongo-driver
```
A good test for me. 😄 The branch is based on the old master（a week ago）

After rebasing upstream/master, I ran the script to update it.

```
➜  kubernetes git:(track-unwanted-dependencies) hack/update-unwanted-dependencies.sh
Move github.com/go-openapi/analysis to not_allowed_dependencies

Move github.com/go-openapi/spec to not_allowed_dependencies

Move github.com/go-openapi/strfmt to not_allowed_dependencies

Move github.com/go-openapi/validate to not_allowed_dependencies

Move go.mongodb.org/mongo-driver to not_allowed_dependencies
```

pacoxu commented on hack/update-unwanted-dependencies.sh:
```diff
@@ -0,0 +1,105 @@
+  not_allowed_dependencies+=("$line")
+done < <(cat "${not_allowed_dependencies_file}")
+
+# Step 1: make sure that not allowed dependencies are not included
```
- Yes, dup script here. The Step 1 is also in `hack/lint-dependencies.sh`.

My concern would be should we have a separate script in this `hack/update-unwanted-dependencies.sh`? 
Can we do the update step in  `hack/lint-dependencies.sh` as well?
Or is update/verify style necessary here?

I may remove it later. I will wait for some more comments on the script structure and fix it together.

gautierdelorme replied:
I think what we should do is:

1. Rename `hack/.not_allowed_dependencies` to `hack/.forbidden_dependencies`
1. In `hack/lint-dependencies.sh` before checking for forbidden dependencies add the logic to verify `hack/.unwanted_dependencies` is up to date with `go mod graph` output.
   1. If it's not up-to-date (i.e. some dependencies were removed) then fail and print in the logs `If you're seeing this locally, run the below command: hack/update-unwanted-dependencies.sh`
   1. If it's up-to-date then do nothing and go to the next step (verify  forbidden dependencies)
1. Then modify `hack/update-unwanted-dependencies.sh` to be an idempotent script that only updates both `hack/.forbidden_dependencies` and `hack/.unwanted_dependencies` to reflect the current state of the project.

So basically in CI we'll only run `hack/lint-dependencies.sh` (same as today) to make sure all dependencies related files are correct and if `hack/.unwanted_dependencies` isn't up-to-date the check will fail and the user will be invited to run `hack/update-unwanted-dependencies.sh` to fix it.

pacoxu replied:
make sense to me, I will try to update it later.

pacoxu commented:
@gautierdelorme  Updated. Thanks for your comments.

gautierdelorme commented on hack/update-unwanted-dependencies.sh:
```diff
@@ -0,0 +1,91 @@
+unwanted_dependencies=()
+while IFS='' read -r line; do
+  unwanted_dependencies+=("$line")
+done < <(cat "${unwanted_dependencies_file}")
+
+# a flag to know if need updated
+changed=false
+
+new_unwanted_dependencies=()
+for unwanted_repo in "${unwanted_dependencies[@]}"; do
+  deps_on_unwanted_repo=$(go mod graph | grep " ${unwanted_repo}@" || echo "")
+  if [ -n "${deps_on_unwanted_repo}" ]; then
+  	# still exist
+  	new_unwanted_dependencies+=("${unwanted_repo}")
+  else
+  	changed=true
+	  not_allowed_dependencies+=("${unwanted_repo}")
+    echo "Move ${unwanted_repo} to not_allowed_dependencies"
+    echo ""
+  fi
+done
+
+if [ "${changed}" = true ]; then
+  echo "Update unwanted and not allowed dependencies"
+
+  echo -n "" > "${unwanted_dependencies_file}.sorted"
+  for unwanted_repo in "${new_unwanted_dependencies[@]}"; do
+  	echo "${unwanted_repo}" >> "${unwanted_dependencies_file}.sorted"
+  done
+
+  echo -n "" > "${not_allowed_dependencies_file}.sorted"
+  for forbidden_repo in "${not_allowed_dependencies[@]}"; do
+  	echo "${forbidden_repo}" >> "${not_allowed_dependencies_file}.sorted"
+  done
+
+  sort "${unwanted_dependencies_file}.sorted" | uniq > "${unwanted_dependencies_file}"
+  sort "${not_allowed_dependencies_file}.sorted" | uniq > "${not_allowed_dependencies_file}"
+  rm "${unwanted_dependencies_file}.sorted" 
+  rm "${not_allowed_dependencies_file}.sorted" 
+fi
```
I think this whole section can be simplified with a single loop, no intermediate `.sorted` files and no `changed` variable:

```bash
unwanted_dependencies=()
while IFS='' read -r unwanted_repo; do
  deps_on_unwanted_repo=$(go mod graph | grep " ${unwanted_repo}@" || echo "")
  if [ -n "${deps_on_unwanted_repo}" ]; then
    unwanted_dependencies+=("${unwanted_repo}")
  else
    forbidden_dependencies+=("${unwanted_repo}")
    echo "Moved ${unwanted_repo} to .forbidden_dependencies"
  fi
done < <(cat "${unwanted_dependencies_file}")

echo "${unwanted_dependencies[*]}" | tr " " "\n" | sort | uniq > "${unwanted_dependencies_file}"
echo "${forbidden_dependencies[*]}" | tr " " "\n" | sort | uniq > "${not_allowed_dependencies_file}"
```

pacoxu replied:
Good point.

BenTheElder reviewed (commented):
can we perhaps document _why_ these are forbidden / unwanted?

maybe we can have `# comment` lines, or at least leave some notes in this PR (for future PRs changing these presumably I can track the git blame to the PR and see the discussion there)

pacoxu commented:
@gautierdelorme I updated the PR. Thanks for you comments.

pacoxu commented:
> can we perhaps document _why_ these are forbidden / unwanted?
> 
> maybe we can have `# comment` lines, or at least leave some notes in this PR (for future PRs changing these presumably I can track the git blame to the PR and see the discussion there)

Good idea.  We may choose a comment style or data structure and continue.

```
# we have switched to klog v2, so avoid klog v1
"k8s.io/klog" 
```
or

```
  "k8s.io/klog"  # we have switched to klog v2, so avoid klog v1
```
or

```
{
  "k8s.io/klog": "we have switched to klog v2, so avoid klog v1"
}
```

Currently, only `github.com/go-kit/kit` is unwanted and a forbidden list like 
- github.com/go-openapi/analysis
- github.com/go-openapi/spec
- github.com/go-openapi/strfmt
- github.com/go-openapi/validate
- github.com/influxdata/influxdb1-client
- go.mongodb.org/mongo-driver
- k8s.io/klog

liggitt commented:
agree on having a comment/reason

I wonder if structured input and writing the check in a real language like go would be preferable (for example, see cmd/importverifier/importverifier.go and the associated .import-restriction input files)

liggitt commented:
inputs could be the file containing unwanted/forbidden inputs, and a file containing the output of `go mod graph`

the tool could then evaluate the `go mod graph` output against the unwanted/forbidden inputs

that structure would allow easy unit testing of hypothetical unwanted/forbidden config vs hypothetical `go mod graph` content

BenTheElder commented:
we might also want to consider making it possible to run some of these dep tools we're writing as one program that can invoke go mod graph once, IIRC it's pretty slow to do this on a repo this size.

liggitt commented:
we're building up quite a list of unwanted deps in https://github.com/kubernetes/kubernetes/issues/102145 ... would be good to get an initial version of this in.

pacoxu commented:
ok， I will work on it this week

pacoxu committed:
add update-unwanted-dependencies.sh to track unwanted dependencies

Signed-off-by: pacoxu <paco.xu@daocloud.io>
Co-authored-by: Jordan Liggitt <jordan@liggitt.net>
Signed-off-by: pacoxu <paco.xu@daocloud.io>

pacoxu commented:
> we're building up quite a list of unwanted deps in #102145 ... would be good to get an initial version of this in.

I updated it and we can move `rsc.io/quote/v3 & rsc.io/sampler` to the forbidden list once #102913 is merged

liggitt commented on hack/lint-dependencies.sh:
```diff
@@ -41,19 +44,22 @@ kube::util::require-jq
-  fi
-done
+# Check if unwanted dependencies are removed
+"${dependencyverifier}" "${KUBE_ROOT}/hack/unwanted-dependencies.json"
```
suggest simplifying this to

```
go run k8s.io/kubernetes/cmd/dependencyverifier "${KUBE_ROOT}/hack/unwanted-dependencies.json"
```

then you don't need to make, find-binary, and verify it was found

liggitt replied:
this is really noisy in the success case... running this, I get the following:

> `$ hack/lint-dependencies.sh `
> `Mod Graph: `
> ` -- map[cloud.google.com/go:cloud.google.com/go/datastore cloud.google.com/go/bigquery:cloud.google.com/go cloud.google.com/go/datastore:cloud.google.com/go cloud.google.com/go/firestore:google.golang.org/grpc cloud.google.com/go/pubsub:cloud.google.com/go cloud.google.com/go/storage:google.golang.org/grpc github.com/Azure/go-autorest/autorest:github.com/Azure/go-autorest/autorest/adal github.com/Azure/go-autorest/autorest/adal:github.com/Azure/go-autorest github.com/Azure/go-autorest/autorest/date:github.com/Azure/go-autorest github.com/Azure/go-autorest/autorest/mocks:github.com/Azure/go-autorest github.com/Azure/go-autorest/autorest/validation:github.com/stretchr/testify github.com/Azure/go-autorest/logger:github.com/Azure/go-autorest github.com/Azure/go-autorest/tracing:github.com/Azure/go-autorest github.com/GoogleCloudPlatform/k8s-cloud-provider:cloud.google.com/go github.com/Microsoft/go-winio:github.com/pkg/errors github.com/Microsoft/hcsshim:github.com/Microsoft/go-winio github.com/NYTimes/gziphandler:github.com/stretchr/testify github.com/alecthomas/units:github.com/stretchr/testify github.com/aws/aws-sdk-go:github.com/jmespath/go-jmespath github.com/bketelsen/crypt:google.golang.org/grpc github.com/checkpoint-restore/go-criu/v5:google.golang.org/protobuf github.com/cilium/ebpf:golang.org/x/sys github.com/cncf/udpa/go:google.golang.org/grpc github.com/cockroachdb/datadriven:github.com/certifi/gocertifi github.com/containerd/cgroups:golang.org/x/sys github.com/containerd/console:golang.org/x/sys github.com/containerd/ttrpc:google.golang.org/grpc github.com/containerd/typeurl:github.com/pkg/errors github.com/coredns/caddy:github.com/google/uuid github.com/coredns/corefile-migration:github.com/coredns/caddy github.com/coreos/go-systemd/v22:github.com/godbus/dbus/v5 github.com/cpuguy83/go-md2man/v2:github.com/shurcooL/sanitized_anchor_name github.com/envoyproxy/go-control-plane:google.golang.org/protobuf github.com/frankban/quicktest:github.com/kr/pretty github.com/fsnotify/fsnotify:golang.org/x/sys github.com/go-kit/log:github.com/go-stack/stack github.com/go-openapi/jsonpointer:github.com/stretchr/testify github.com/go-openapi/jsonreference:golang.org/x/text github.com/go-openapi/swag:gopkg.in/yaml.v2 github.com/gogo/protobuf:golang.org/x/tools github.com/golang/mock:rsc.io/quote/v3 github.com/golang/protobuf:google.golang.org/protobuf github.com/google/cadvisor:cloud.google.com/go github.com/google/go-cmp:golang.org/x/xerrors github.com/google/pprof:golang.org/x/sys github.com/googleapis/gax-go/v2:google.golang.org/grpc github.com/googleapis/gnostic:github.com/docopt/docopt-go github.com/gophercloud/gophercloud:gopkg.in/yaml.v2 github.com/grpc-ecosystem/go-grpc-middleware:google.golang.org/grpc github.com/grpc-ecosystem/grpc-gateway:gopkg.in/yaml.v2 github.com/hashicorp/consul/api:github.com/hashicorp/consul/sdk github.com/hashicorp/consul/sdk:github.com/hashicorp/go-cleanhttp github.com/hashicorp/go-immutable-radix:github.com/hashicorp/go-uuid github.com/hashicorp/go-multierror:github.com/hashicorp/errwrap github.com/hashicorp/go-rootcerts:github.com/mitchellh/go-homedir github.com/hashicorp/hcl:github.com/davecgh/go-spew github.com/hashicorp/mdns:github.com/hashicorp/go.net github.com/hashicorp/memberlist:github.com/armon/go-metrics github.com/hashicorp/serf:github.com/armon/circbuf github.com/jmespath/go-jmespath:github.com/jmespath/go-jmespath/internal/testify github.com/jmespath/go-jmespath/internal/testify:gopkg.in/yaml.v2 github.com/json-iterator/go:github.com/stretchr/testify github.com/kisielk/errcheck:golang.org/x/tools github.com/kr/pretty:github.com/kr/text github.com/kr/text:github.com/creack/pty github.com/matttproud/golang_protobuf_extensions:golang.org/x/sync github.com/miekg/dns:golang.org/x/crypto github.com/mitchellh/cli:github.com/armon/go-radix github.com/moby/ipvs:github.com/pkg/errors github.com/moby/spdystream:github.com/gorilla/websocket github.com/moby/sys/mountinfo:golang.org/x/sys github.com/moby/term:gotest.tools/v3 github.com/niemeyer/pretty:github.com/kr/text github.com/nxadm/tail:gopkg.in/tomb.v1 github.com/olekukonko/tablewriter:github.com/mattn/go-runewidth github.com/onsi/ginkgo:golang.org/x/text github.com/onsi/gomega:gopkg.in/yaml.v2 github.com/opencontainers/runc:github.com/checkpoint-restore/go-criu/v5 github.com/opencontainers/selinux:github.com/pkg/errors github.com/prometheus/client_golang:google.golang.org/protobuf github.com/prometheus/client_model:golang.org/x/sync github.com/prometheus/common:gopkg.in/yaml.v2 github.com/prometheus/procfs:golang.org/x/sys github.com/rogpeppe/go-internal:gopkg.in/errgo.v2 github.com/sergi/go-diff:github.com/davecgh/go-spew github.com/sirupsen/logrus:golang.org/x/sys github.com/smartystreets/goconvey:golang.org/x/tools github.com/soheilhy/cmux:golang.org/x/net github.com/spf13/afero:golang.org/x/text github.com/spf13/cast:github.com/stretchr/testify github.com/spf13/cobra:gopkg.in/yaml.v2 github.com/spf13/viper:gopkg.in/yaml.v2 github.com/stoewer/go-strcase:github.com/stretchr/testify github.com/stretchr/objx:github.com/stretchr/testify github.com/stretchr/testify:gopkg.in/yaml.v3 github.com/urfave/cli:gopkg.in/yaml.v2 github.com/vishvananda/netlink:github.com/vishvananda/netns github.com/vishvananda/netns:golang.org/x/sys go.etcd.io/bbolt:golang.org/x/sys go.etcd.io/etcd/api/v3:google.golang.org/grpc go.etcd.io/etcd/client/pkg/v3:github.com/coreos/go-systemd/v22 go.etcd.io/etcd/client/v2:go.etcd.io/etcd/client/pkg/v3 go.etcd.io/etcd/client/v3:github.com/dustin/go-humanize go.etcd.io/etcd/etcdutl/v3:go.uber.org/zap go.etcd.io/etcd/pkg/v3:google.golang.org/grpc go.etcd.io/etcd/raft/v3:go.etcd.io/etcd/client/pkg/v3 go.etcd.io/etcd/server/v3:sigs.k8s.io/yaml go.etcd.io/etcd/tests/v3:gopkg.in/yaml.v2 go.opencensus.io:google.golang.org/grpc go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc:github.com/golang/protobuf go.opentelemetry.io/otel:github.com/google/go-cmp go.opentelemetry.io/otel/exporters/otlp:github.com/google/go-cmp go.opentelemetry.io/otel/metric:github.com/google/go-cmp go.opentelemetry.io/otel/oteltest:go.opentelemetry.io/otel go.opentelemetry.io/otel/sdk:github.com/google/go-cmp go.opentelemetry.io/otel/sdk/export/metric:github.com/stretchr/testify go.opentelemetry.io/otel/sdk/metric:github.com/benbjohnson/clock go.opentelemetry.io/otel/trace:github.com/google/go-cmp go.opentelemetry.io/proto/otlp:github.com/golang/protobuf go.starlark.net:github.com/chzyer/logex go.uber.org/atomic:github.com/stretchr/testify go.uber.org/goleak:github.com/kr/pretty go.uber.org/multierr:go.uber.org/atomic go.uber.org/zap:gopkg.in/yaml.v3 golang.org/x/crypto:golang.org/x/term golang.org/x/exp:golang.org/x/xerrors golang.org/x/image:golang.org/x/text golang.org/x/lint:golang.org/x/tools golang.org/x/mobile:golang.org/x/tools golang.org/x/mod:golang.org/x/xerrors golang.org/x/net:golang.org/x/text golang.org/x/oauth2:google.golang.org/appengine golang.org/x/term:golang.org/x/sys golang.org/x/text:golang.org/x/tools golang.org/x/tools:golang.org/x/xerrors gonum.org/v1/gonum:gonum.org/v1/plot gonum.org/v1/netlib:modernc.org/xc gonum.org/v1/plot:rsc.io/pdf google.golang.org/api:cloud.google.com/go google.golang.org/appengine:golang.org/x/text google.golang.org/genproto:google.golang.org/protobuf google.golang.org/grpc:google.golang.org/protobuf google.golang.org/protobuf:github.com/google/go-cmp gopkg.in/errgo.v2:github.com/kr/pretty gopkg.in/yaml.v2:gopkg.in/check.v1 gopkg.in/yaml.v3:gopkg.in/check.v1 gotest.tools/v3:golang.org/x/tools honnef.co/go/tools:golang.org/x/mod k8s.io/api:github.com/gogo/protobuf k8s.io/apiextensions-apiserver:github.com/emicklei/go-restful k8s.io/apimachinery:sigs.k8s.io/yaml k8s.io/apiserver:github.com/google/uuid k8s.io/cli-runtime:github.com/davecgh/go-spew k8s.io/client-go:cloud.google.com/go k8s.io/cloud-provider:github.com/google/go-cmp k8s.io/cluster-bootstrap:github.com/stretchr/testify k8s.io/code-generator:github.com/emicklei/go-restful k8s.io/component-base:github.com/blang/semver k8s.io/component-helpers:github.com/google/go-cmp k8s.io/controller-manager:github.com/spf13/pflag k8s.io/cri-api:github.com/davecgh/go-spew k8s.io/csi-translation-lib:github.com/stretchr/testify k8s.io/gengo:github.com/davecgh/go-spew k8s.io/klog/v2:github.com/go-logr/logr k8s.io/kube-aggregator:github.com/spf13/pflag k8s.io/kube-controller-manager:k8s.io/apimachinery k8s.io/kube-openapi:github.com/NYTimes/gziphandler k8s.io/kube-proxy:k8s.io/apimachinery k8s.io/kube-scheduler:github.com/google/go-cmp k8s.io/kubectl:github.com/MakeNowJust/heredoc k8s.io/kubelet:github.com/gogo/protobuf k8s.io/kubernetes:sigs.k8s.io/yaml k8s.io/legacy-cloud-providers:cloud.google.com/go k8s.io/metrics:github.com/gogo/protobuf k8s.io/mount-utils:github.com/kr/text k8s.io/sample-apiserver:github.com/google/gofuzz k8s.io/system-validators:github.com/blang/semver k8s.io/utils:github.com/davecgh/go-spew rsc.io/quote/v3:rsc.io/sampler rsc.io/sampler:golang.org/x/text sigs.k8s.io/apiserver-network-proxy/konnectivity-client:k8s.io/klog/v2 sigs.k8s.io/kustomize/api:k8s.io/kube-openapi sigs.k8s.io/kustomize/cmd/config:k8s.io/kube-openapi sigs.k8s.io/kustomize/kustomize/v4:sigs.k8s.io/yaml sigs.k8s.io/kustomize/kyaml:k8s.io/kube-openapi sigs.k8s.io/structured-merge-diff/v4:gopkg.in/yaml.v2 sigs.k8s.io/yaml:gopkg.in/yaml.v2]`
> `All dependencies:`
> ` -- { `
> `  "forbidden-dependencies": {`
> `    "github.com/go-kit/kit": "",`
> `    "github.com/go-openapi/analysis": "",`
> `    "github.com/go-openapi/spec": "",`
> `    "github.com/go-openapi/strfmt": "",`
> `    "github.com/go-openapi/validate": "",`
> `    "github.com/influxdata/influxdb1-client": "",`
> `    "go.mongodb.org/mongo-driver": "",`
> `    "k8s.io/klog": "we have switched to klog v2, so avoid klog v1",`
> `    "rsc.io/quote": "refer to #102833"`
> `  },`
> `  "unwanted-dependencies": {`
> `    "github.com/hashicorp/hcl": "",`
> `    "github.com/spf13/viper": "refer to #102598",`
> `    "rsc.io/sampler": "refer to #102833"`
> `  }`
> `}`
> 
> `Forbidden dependencies:`
> ` -- map[github.com/go-kit/kit: github.com/go-openapi/analysis: github.com/go-openapi/spec: github.com/go-openapi/strfmt: github.com/go-openapi/validate: github.com/influxdata/influxdb1-client: go.mongodb.org/mongo-driver: k8s.io/klog:we have switched to klog v2, so avoid klog v1 rsc.io/quote:refer to #102833]`
> `Unwanted dependencies:`
> ` -- map[github.com/hashicorp/hcl: github.com/spf13/viper:refer to #102598 rsc.io/sampler:refer to #102833]`
> `2021/06/29 14:10:14 - Check forbidden depencies rsc.io/quote...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/go-kit/kit...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/go-openapi/spec...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/influxdata/influxdb1-client...`
> `2021/06/29 14:10:14 - Check forbidden depencies k8s.io/klog...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/go-openapi/analysis...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/go-openapi/strfmt...`
> `2021/06/29 14:10:14 - Check forbidden depencies github.com/go-openapi/validate...`
> `2021/06/29 14:10:14 - Check forbidden depencies go.mongodb.org/mongo-driver...`
> `2021/06/29 14:10:14 - Check unwanted depencies github.com/hashicorp/hcl...`
> `2021/06/29 14:10:14 - Check unwanted depencies github.com/spf13/viper...`
> `2021/06/29 14:10:14 - Check unwanted depencies rsc.io/sampler...`
> 
> `real	0m13.041s`
> `user	0m1.134s`
> `sys	0m2.557s`
> `All pinned versions of checked dependencies match their preferred version.`
> 
> `$ echo $?`
> `0`

That looks a lot like an error, not a successful run

pacoxu replied:
I removed some debug messages.

Updated.

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -0,0 +1,18 @@
+{ 
+  "forbidden-dependencies": {
+    "github.com/go-kit/kit": "",
```
```suggestion
    "github.com/go-kit/kit": "lots of transitive dependencies, see https://github.com/prometheus/common/issues/255",
```

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -0,0 +1,18 @@
+{ 
+  "forbidden-dependencies": {
+    "github.com/go-kit/kit": "",
+    "github.com/go-openapi/analysis": "",
+    "github.com/go-openapi/spec": "",
+    "github.com/go-openapi/strfmt": "",
+    "github.com/go-openapi/validate": "",
```
```suggestion
    "github.com/go-openapi/analysis": "use k8s.io/kube-openapi/pkg/validation/spec",
    "github.com/go-openapi/spec": "use k8s.io/kube-openapi/pkg/validation/spec instead",
    "github.com/go-openapi/strfmt": "use k8s.io/kube-openapi/pkg/validation/strfmt instead",
    "github.com/go-openapi/validate": "use k8s.io/kube-openapi/pkg/validation/validate instead",
```

liggitt commented on hack/lint-dependencies.sh:
```diff
@@ -25,6 +25,9 @@ set -o pipefail
 KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 source "${KUBE_ROOT}/hack/lib/init.sh"
 
+kube::golang::setup_env
+make -C "${KUBE_ROOT}" WHAT=cmd/dependencyverifier
```
can drop this, we're running directly below

pacoxu replied:
Removed.

liggitt commented on cmd/dependencyverifier/dependencyverifier.go:
```diff
@@ -0,0 +1,133 @@
+	return dependencies[key]
+}
+
+func convertToMap(modStr string) map[string]string {
```
this doesn't look quite right... the left side is the parent, the right side is the dependent... returning map[string]string means that we can only express a single parent or a single child for a given module

I think we want to return `map[string][]string`, where the key is the child module (the right side of `go mod graph` output) and the value is the list of module names that refer to it

pacoxu replied:
Updated.

liggitt commented on cmd/dependencyverifier/dependencyverifier.go:
```diff
@@ -0,0 +1,133 @@
+	"os/exec"
+	"strings"
+)
+
```
thinking through the workflow of identifying a module we don't want, then gradually eliminating it from being vendored, from referencing it directly, and from referencing it indirectly, something like this might be clearer as input config:

```go
type Unwanted struct {
	// things we want to stop referencing
	Spec UnwantedSpec `json:"spec"`
	// status of our unwanted dependencies
	Status UnwantedStatus `json:"status"`
}
type UnwantedSpec struct {
	// module names/patterns of root modules whose dependencies should be considered direct references
	RootModules []string  `json:"rootModules"`
	// module names we don't want to depend on, mapped to an optional message about why
	UnwantedModules map[string]string  `json:"unwantedModules"`
}
type UnwantedStatus struct {
	// unwanted modules we still vendor, based on vendor/modules.txt content.
	// eliminating things from this list is good.
	Vendored []string  `json:"vendored"`
	// unwanted modules we still directly reference from spec.roots, based on `go mod graph` content.
	// eliminating things from this list is good.
	DirectReferences []string  `json:"directReferences"`
	// unwanted modules indirectly referenced from modules other than spec.roots, based on `go mod graph` content.
	// eliminating things from this list is good, but usually requires working with upstreams to do so.
	IndirectReferences []string `json:"indirectReferences"`
}
```

you can get syntax checking for free from the json decoder like this:
```
       config := &Unwanted{}
       decoder:=json.NewDecoder(bytes.NewBuffer([]byte(dependencies)))
       decoder.DisallowUnknownFields()
       if err := decoder.Decode(config);err != nil {
               log.Fatalf("Error reading dependencies file %s: %s", dependenciesJSONPath, err)
       }
```

then we cleanly separate intent ("here are our root modules, here are our unwanted modules") from status ("still vendored, still referenced from root modules, still referenced indirectly")

our input file would look like:
```yaml
{ 
  "spec": {
    "rootModules": [
      "k8s.io/*"
    ],
    "unwantedModules": {
      "github.com/go-kit/kit": "lots of transitive dependencies, see https://github.com/prometheus/common/issues/255",
      "github.com/go-openapi/analysis": "use k8s.io/kube-openapi/pkg/validation/spec",
      ...
      "github.com/hashicorp/hcl": "",
      "github.com/spf13/viper": "refer to #102598"
    }
  },
  "status": {
    "vendored": [
      ... unwantedModules still showing as vendored in modules.txt
    ],
    "directReferences": [
      ... unwantedModules still showing as dependencies of spec.rootModules in go mod graph output
    ],
    "indirectReferences": [
      ... unwantedModules still showing as dependencies of modules other than spec.rootModules in go mod graph output
    ]
  }
}
```

Then the tool can compute status, do a deepequal check against the input's status and if they differ, dump the expected status

If we start in this direction in this PR just with spec.unwantedModules / status.indirectReferences, the rootModules/vendored/directReferences bits can be added in follow-ups

pacoxu replied:
Updated with the structs and currently only update `status.indirectReferences` according to `UnwantedModules`.

pacoxu committed:
use reference as we cannot distinguishing direct/indirect with `go mod graph`

pacoxu commented:
Update the unwanted list
- github.com/go-bindata/go-bindata (Replace go-bindata with //go:embed #99829) unwanted
- github.com/hashicorp/golang-lru not allowed

liggitt commented on cmd/dependencyverifier/dependencyverifier.go:
```diff
@@ -0,0 +1,181 @@
+	// eliminating things from this list is good.
+	DirectReferences []string `json:"directReferences"`
+
+	// unwanted modules indirectly referenced from modules other than spec.roots, based on `go mod graph` content.
```
Feedback from the go team and behavior change in go1.17 indicate `go mod graph` isn't the right tool to determine direct/indirect dependencies (https://github.com/kubernetes/kubernetes/issues/104461). However, since `go list` does not return results for all build tags, there's not a clear alternative at the moment.

dims replied:
Ugh, so how can we make progress here? :(

liggitt replied:
hoisting from slack:

```
we can use go mod graph to get the full set of dependencies
which is what the tool is using now
but to resolve the "TODOs" distinguishing direct/indirect, it won't help us
so it would still be useful for keeping an unwanted dependency we've eradicated from creeping back in
but it won't be useful for keeping one that is currently an indirect dep from creeping in as a direct dep (edited) 
```

dims commented:
/assign

dims commented:
i am happy with this. 

/approve

will wait for @pacoxu to peek at the last comments here and on slack before we ship this.

pacoxu commented:
Updated. 

1. Add `github.com/json-iterator/go` as unwanted.
2. As we cannot distinguishing direct/indirect, use simple `reference` in the struct. And add `TODO` to distinguish them.

dims commented:
/assign @thockin @liggitt 

(for `/cmd` approval)

liggitt commented on cmd/dependencyverifier/OWNERS:
```diff
@@ -0,0 +1,9 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+reviewers:
```
also add `dep-reviewers` here

liggitt commented on cmd/dependencyverifier/OWNERS:
```diff
@@ -0,0 +1,9 @@
+approvers:
+  - bentheelder
+  - pacoxu
+  - liggitt
```
drop ben and I and add `dep-approvers` instead

liggitt commented:
this is the current output if a module adds a bad dependency on klog:
```
!!! [0917 10:35:56] The following have transitive dependencies on k8s.io/klog, which is not allowed:
k8s.io/sample-apiserver@v0.0.0 k8s.io/klog@v1.0.0
```

with this PR, the output changes to:
```
Status in ./hack/unwanted-dependencies.json:
 {[] [github.com/go-kit/kit github.com/hashicorp/golang-lru github.com/hashicorp/hcl github.com/json-iterator/go github.com/spf13/viper]}
Status detected:
 {[] [github.com/go-kit/kit github.com/hashicorp/golang-lru github.com/hashicorp/hcl github.com/json-iterator/go github.com/spf13/viper k8s.io/klog]}
2021/09/17 10:34:21 !!! Please update ./hack/unwanted-dependencies.json
```

A couple observations:
1. it's hard to tell which module has the bad dependency
2. when there's a delta in the unwanted status that _reintroduces_ an unwanted dependency, we _don't_ actually want them to update `hack/unwanted-dependencies.json` to allow it (which is what the message encourages them to do)

I think this needs to be clearer when there's a mismatch in the status. For anything currently in the file that should be removed from status, that's great and we can just tell them to drop the thing from status. For anything not in the file that is newly referenced, we should print out the modules depending on the unwanted dependency so they can fix those.

pacoxu commented:
I just try to refactor the output. And I will update the `func diff` which is not that graceful tommorrow.

```
➜  kubernetes git:(track-unwanted-dependencies-1) ✗ hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
All pinned versions of checked dependencies match their preferred version.
```

For a removed dependency
```
➜  kubernetes git:(track-unwanted-dependencies-1) ✗ hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
2021/09/17 23:35:46 The following dependencies are removed:
k8s.io/klog
exit status 1
```

A dependency `k8s.io/klog/v2` which is not allowed, is used .
```
➜  kubernetes git:(track-unwanted-dependencies-1) ✗ hack/lint-dependencies.sh
Check all unwanted dependencies and update its status.
2021/09/17 23:36:13 The following dependencies are not allowed:
k8s.io/klog/v2
exit status 1
```

liggitt commented:
This looks closer, but I think we still need to be more specific about which modules are referencing the unwanted dependencies, and what action should be taken. Something like the following gives different guidance depending on which situation the user encounters:

```diff
diff --git a/cmd/dependencyverifier/dependencyverifier.go b/cmd/dependencyverifier/dependencyverifier.go
index 26f56952996..ecf367945a3 100644
--- a/cmd/dependencyverifier/dependencyverifier.go
+++ b/cmd/dependencyverifier/dependencyverifier.go
@@ -187,21 +187,22 @@ func main() {
 	// Compare unwanted list from unwanted-dependencies.json with current status from `go mod graph`
 	removedReferences, unwantedReferences := difference(configFromFile.Status.References, config.Status.References)
 	if len(removedReferences) > 0 {
-		log.Println("The following dependencies are removed:")
+		log.Println("Good news! The following unwanted dependencies are no longer referenced:")
 		for reference := range removedReferences {
-			fmt.Println(reference)
+			log.Printf("   %s", reference)
 		}
+		log.Printf("!!! Remove the unwanted dependencies from status in %s to ensure they don't get reintroduced", dependenciesJSONPath)
 		needUpdate = true
 	}
 	if len(unwantedReferences) > 0 {
-		log.Println("The following dependencies are not allowed:")
+		log.Printf("The following unwanted dependencies marked in %s are referenced:", dependenciesJSONPath)
 		for reference := range unwantedReferences {
-			fmt.Println(reference)
+			log.Printf("   %s (referenced by %s)", reference, strings.Join(modeGraph[reference], ", "))
 		}
+		log.Printf("!!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies\n")
 		needUpdate = true
 	}
 	if needUpdate {
-		log.Println("!!! Please update status in ./hack/unwanted-dependencies.json")
 		os.Exit(1)
 	}
 }
```

pacoxu commented:
Updated, new output is like below.

```
2021/09/20 22:15:11 Good news! The following unwanted dependencies are no longer referenced:
2021/09/20 22:15:11    k8s.io/klog
2021/09/20 22:15:11 !!! Remove the unwanted dependencies from status in /Users/pacoxu/git/gopath/src/k8s.io/kubernetes/hack/unwanted-dependencies.json to ensure they don't get reintroduced

2021/09/20 22:15:11 The following unwanted dependencies marked in /Users/pacoxu/git/gopath/src/k8s.io/kubernetes/hack/unwanted-dependencies.json are referenced:
2021/09/20 22:15:11    github.com/go-kit/kit (referenced by github.com/grpc-ecosystem/go-grpc-middleware)
2021/09/20 22:15:11 !!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies
exit status 1
```

liggitt commented:
/lgtm
/approve
